ImageFeed/Authorization/Services/OAuth2Service.swift

```diff
@@ -12,77 +12,59 @@ final class OAuth2Service {
         case codeError
     }
 
+    private var task: URLSessionTask?
+    private var lastCode: String?
+
+    private let session = URLSession.shared
+    private let network = Network.shared
+
     func fetchAuthToken(
         code: String,
         completition: @escaping (Result<String, Error>) -> Void
     ) {
-        guard let urlComponents = URLComponents(
-            string: API.OAuthTokenURLString
-        ) else {
-            assertionFailure("Auth token URL is unvailable")
-            return
-        }
+        assert(Thread.isMainThread)
 
-        var composedURL = urlComponents
-        composedURL.queryItems = [
-            URLQueryItem(
-                name: "client_id",
-                value: API.accessKey
-            ),
-            URLQueryItem(
-                name: "client_secret",
-                value: API.secretKey
-            ),
-            URLQueryItem(
-                name: "redirect_uri",
-                value: API.redirectURI
-            ),
-            URLQueryItem(
-                name: "code",
-                value: code
-            ),
-            URLQueryItem(
-                name: "grant_type",
-                value: "authorization_code"
-            )
-        ]
+        if lastCode == code { return }
+        task?.cancel()
+        lastCode = code
 
-        guard let url = composedURL.url else {
-            assertionFailure("Unable to construct composed Auth token URL")
-            return
-        }
+        let request = makeAuthTokenRequest(code: code)
 
-        var request = URLRequest(url: url)
-        request.httpMethod = "POST"
+        let task = session.objectTask(for: request) {
+            [weak self] (result: Result<OAuthToken, Error>) in
+            guard let self else { return }
```

Comment:
список захвата ([weak self] и вычисляемый тип нужно оставлять на той же строке, где открылась фигурная скобка
Вот так общеупотребимо
```suggestion
        let task = session.objectTask(for: request) { [weak self] (result: Result<OAuthToken, Error>) in
            guard let self else { return }
```

Reply:
fixed